inputgraph/loader.go

```diff
@@ -26,76 +27,89 @@ var (
 	ErrUnableToDetermineHash = fmt.Errorf("unable to determine hash")
 )
 
-func argsContainsStr(args []string, substr string) bool {
-	for _, s := range args {
-		if strings.Contains(s, substr) {
-			return true
-		}
-	}
-	return false
-}
-
-func requiresCrossProduct(args []string) bool {
-	seen := map[string]struct{}{}
-	for _, s := range args {
-		k := strings.SplitN(s, "=", 2)[0]
-		if _, found := seen[k]; found {
-			return true
-		}
-		seen[k] = struct{}{}
+type loader struct {
+	conslog          conslogging.ConsoleLogger
+	target           domain.Target
+	visited          map[string]struct{}
+	hasher           *hasher.Hasher
+	varCollection    *variables.Collection
+	features         *features.Features
+	isBaseTarget     bool
+	ci               bool
+	push             bool
+	builtinArgs      variables.DefaultArgs
+	overridingVars   *variables.Scope
+	earthlyCIRunner  bool
+	platformResolver *platutil.Resolver
+}
+
+func newLoader(ctx context.Context, opt HashOpt) *loader {
+	// Other important values are set by load().
+	return &loader{
+		conslog:          opt.Console,
+		target:           opt.Target,
+		visited:          map[string]struct{}{},
+		hasher:           hasher.New(),
+		isBaseTarget:     opt.Target.Target == "base",
+		ci:               opt.CI,
+		push:             opt.Push,
+		builtinArgs:      opt.BuiltinArgs,
+		overridingVars:   opt.OverridingVars,
+		earthlyCIRunner:  opt.EarthlyCIRunner,
+		platformResolver: opt.PlatformResolver,
 	}
-	return false
-}
-
-func getArgsCopy(cmd spec.Command) []string {
-	argsCopy := make([]string, len(cmd.Args))
-	copy(argsCopy, cmd.Args)
-	return argsCopy
-}
-
-func parseArgs(cmdName string, opts interface{}, args []string) ([]string, error) {
-	processed := stringutil.ProcessParamsAndQuotes(args)
-	return flagutil.ParseArgs(cmdName, opts, processed)
 }
 
 func (l *loader) handleFrom(ctx context.Context, cmd spec.Command) error {
+	cmd.Args = l.expandArgs(ctx, cmd.Args)
```

Comment:
For the actual argument expansion, I'm using this method for each command. Does this look correct?

Could this approach lead to issues with shell-out behavior?